x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -727,16 +814,313 @@ public void runAsDenied(String requestId, Authentication authentication, RestReq
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "delete_user");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject()
+                    .startObject("user")
+                        .field("name", deleteUserRequest.username())
+                    .endObject() // user
+                    .endObject();
+            logEntry.with(DELETE_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(DeleteRoleRequest deleteRoleRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "delete_role");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject()
+                    .startObject("role")
+                    .field("name", deleteRoleRequest.name())
+                    .endObject() // role
+                    .endObject();
+            logEntry.with(DELETE_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(DeleteRoleMappingRequest deleteRoleMappingRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "delete_role_mapping");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject()
+                    .startObject("role_mapping")
+                    .field("name", deleteRoleMappingRequest.getName())
+                    .endObject() // role_mapping
+                    .endObject();
+            logEntry.with(DELETE_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(InvalidateApiKeyRequest invalidateApiKeyRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "invalidate_apikeys");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject()
+                    .startObject("apikeys");
+                if (invalidateApiKeyRequest.getIds() != null && invalidateApiKeyRequest.getIds().length > 0) {
+                    builder.array("ids", invalidateApiKeyRequest.getIds());
+                }
+                if (Strings.hasLength(invalidateApiKeyRequest.getName())) {
+                    builder.field("name", invalidateApiKeyRequest.getName());
+                }
+                builder.field("owned_by_authenticated_user", invalidateApiKeyRequest.ownedByAuthenticatedUser());
+                if (Strings.hasLength(invalidateApiKeyRequest.getUserName()) || Strings.hasLength(invalidateApiKeyRequest.getRealmName())) {
+                    builder.startObject("user")
+                            .field("name", invalidateApiKeyRequest.getUserName())
+                            .field("realm", invalidateApiKeyRequest.getRealmName())
```

Comment:
It is possible that one of them could be empty. Do we need to skip the empty value?

Reply:
I don't think so (hence the code). For example, in roles, we don't skip empty fields, we output them as `[]` instead of `null` or not showing them at all.
In this case, the values of `realm` and `name` are important when missing, as they're interpreted as `all` or `*`.

Reply:
With the current form, an empty value, let's say an empty `realm` value, can be output as `null` or `""` or `"   "`. I get it that we want to output empty value in this case. But is this inconsitency helpful?

Reply:
In general, the fact that different **internal** requests amount to the same result is not helpful.
More to the point, nulls, empty and whitespace strings are a nuisance, but there are two issues conflated here:
* should we formalize the `InvalidateApiKeyRequest` object, such that a missing `realmName` or `userName` parameter is represented in the same way (equivalently, add methods `isRealmNameSet` that are then called in the action handler, rather than leave the action handler interpret the parameter)? I think we should.
* should we audit the value for `missing` (that we specify in the above)? There are precedents "for" (eg cluster/index/app privileges in a role descriptor) as well as "against" (eg DLS/FLS permissions on an index privilege). My guiding principle is to audit it, unless it is uncommon and in the process creates noise. If you ask me, in this case I would audit, because `null` here really means `all` (it's a special kind of `null` to the other places).

Reply:
I think some of the issue here is the usage of `Strings#hasLength`, while the Request and Handler code uses `Strings#hasText`, which eliminates strings of all whitespace. So if you change it to `hasText`, it will at least get rid of the logging for a string like `"    "`.

I wasn't sure before but now I think I prefer to skip the empty values:
* If `null` values are really that important here and should be logged, then we should always log them. This means even when both `realmName` and `userName` are null, we should log them, because they mean `all`. But I think this is apparently too much. So the opposite approach, which I prefer, is to not log them on missing values.
* Because of the internal leniency of the `InvalidateApiKeyRequest`, empty values can take multiple forms. It is a hassle to consolidate them. And as you said, the right place for the consolidation is in the Request class itself. But I don't think we should drag it into this PR. Therefore, instead of logging different formats of "null" value, skipping them at least has the benefit of consistency.
* Even though `null` has the practical effect of `all` here, I don't think it really is that special. In my opinion, it is quite obvious because all these variables are _search_ criteria. If not specified, it's obvious that it should take no effect. Also strictly speaking, `null` does *not* mean `all`. It really just means "missing" because you will not see it in the final search query at all. The cluster privileges in `RoleDescriptor` is a different story because it is not a search query, so its meaning could be ambiguous when skipped. If anything, the RoleDescriptor's `query` field is skipped on missing value because it is also a search criteria.

So my concrete suggestion would be to use `Strings#hasText` here and skip empty values.

Reply:
I would support cleaning up the request object so that it didn't support different styles of "empty" strings - it may be as simple as changing the setters to have 
```
if( Strings.hasText(arg) ) {
   this.field = arg;
} else 
   this.field = null;
}
```
But, I'm not sure there's a good reason to squeeze that into this PR.

I think the current logging is the preferred behaviour - or at least we should always print them as a pair, I don't have a string view whether we should output the match-all fields as `*`, `_all` or `null`.
The `user` and `realm` are a pair. They work as a pair, and they should be printed as a pair.

Reply:
I opened https://github.com/elastic/elasticsearch/pull/66240 to unify the `hasText` checks.

Reply:
> The `user` and `realm` are a pair. They work as a pair, and they should be printed as a pair.
>
This is a good point. Within the realm system, these two variables are indeed a pair. However, the Invalidate API Key API does not realy treat them as a pair in that 1) you don't need specify both of them; 2) their value absence is handled the same way as for other variables, e.g. API key `name`, i.e. the value absence just means the search critiera does not include this variable. So there is nothing inherently with this particular API that makes the two feel like a pair. For example, you can invalidate all keys belong to user `foo` from multiple realms: 
```javascript
DELETE _security/api_key
{"username":"foo"}
```
Is this behaviour intended?

With above being said, I agree to output username and realm as a pair since this should be the norm. If there needs any fix, I think it is with the InvalidateAPIKey API and its validation logic.

Reply:
> Is this behaviour intended?

It was intentional, though an argument could be made that we should have enforced them being entered as a pair, and explicitly have
```
DELETE _security/api_key
{
  "username":"foo",
  "realm":"*"
}
```

But it was intentional that you should be able to delete for a username regardless of realm (in case you only have the default realms, and don't know/care what they're called).

Likewise you can invalidate for a whole realm, e.g. if you had a security compromise and are concerned that all the passwords for an LDAP realm might have been leaked, but we _could_ have made it explicit and required `"username":"*"`.